### ISSUE TITLE
False positive for error 7022: "referenced directly or indirectly in its own initializer."

### ISSUE THREAD

**Issue by denis-migdal** [microsoft/TypeScript#62394]
### 🔎 Search Terms

referenced directly or indirectly in its own initializer.

### 🕗 Version & Regression Information

Tested in playground on 5.9.2 (nightly isn't loading both in FF and Chromium).

### ⏯ Playground Link

[Playground Link](https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABAWxgDxmAzgCgJSIDeAUImYgE4CmUIFSEANgIZZZGnlcQJZQUhocCjmYAuRAEEChAL6cy8+cR7YoiALKTEAXhTpMuPAG5iKlmykcuzXYjBUA7psk4oACxhYTxZU1bsAEKIVGhQVGAAJuzaJDZ2Ds5abp7eprJAA)

### 💻 Code

```ts
function mixins() {
    return class {
        constructor(a: A) {}
    }
}

const MA = mixins();

class A {
    a = new MA(this);
}

class B extends A {
    a = new MA(this);
}
```

### 🙁 Actual behavior

In class B, we have the following error :
```
'a' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.(7022)
```

### 🙂 Expected behavior

No TS error.

### Additional information about the issue

The error doesn't appear when we have :
```ts
function mixins() {
    return class {
        constructor(a: any) {}
    }
}
```

Possible related issues:
- https://github.com/microsoft/TypeScript/issues/59271
- https://github.com/microsoft/TypeScript/issues/55640
- https://github.com/microsoft/TypeScript/issues/48708

**Comment by RyanCavanaugh** [comment]
This is a true circularity, since we can't short-circuit to an immediate nominal `A === A` check on the call.

**Comment by denis-migdal** [comment]
> This is a true circularity, since we can't short-circuit to an immediate nominal `A === A` check on the call.

Well, we do have a circularity in class A, but TS is clever enough to do its stuff.
Then TS should also be clever enough for class B.

TS should know that B "is a" A, and therefore the return type of MA.

The following works too:
```ts
class MA {
    constructor(a: A) {}
}

class A {
    a = new MA(this);
}

class MB {
    constructor(a: B) {}
}

class B extends A {
    a = new MB(this);
}
```

Works also with generics :
```ts
class MA<T extends any> {
    constructor(a: T) {}
}
```
(doesn't work if we extends A though).

**Comment by RyanCavanaugh** [comment]
> TS should know that B "is a" A

*How* does it evaluate this? By seeing if the properties of `B` are assignable to `A`. How does it see what the properties are? By evaluating the initializers. What's the initializer doing? Trying to use a `B` as an `A`

See also https://github.com/Microsoft/TypeScript/wiki/FAQ#circularity-errors-may-occur-in-the-presence-of-circularities

**Comment by denis-migdal** [comment]
And were is the circularity ?

The hypothesis to verify is : "B properly extends A".
Being able  to "use a B as an A" is a consequence of the hypothesis, it will be true if the hypothesis is true... so we can continue...
This is false only if the hypothesis is false, i.e. if we we find an error somewhere else... and if we do so, then we'll have an error somewhere else to repport.
When testing the hypothesis, we can safely assume the hypothesis is true for this kind of case.

Please note that `MA` isn't generic, we know its return type whatever the arguments.
Meaning that we don't need to verify the call to get its return type, and simply verifying it after verifying that "B properly extends A".

**Comment by RyanCavanaugh** [comment]
This is addressed in the FAQ.

**Comment by denis-migdal** [comment]
So, this isn't a defect until someone fix it...

Pretty sure you could solve lot of things with some graph theory.
